feat(diff): add DiffTool command for file and directory comparison

deathbeam opened the pull request:
Separate repo for testing this plugin is here: https://github.com/deathbeam/difftool.nvim without requiring nvim rebuild

Problem:
Neovim's built-in diff mode (nvim -d) does not support directory diffing
as required by git difftool -d. This makes it difficult to compare
entire directories, detect renames, and navigate changes efficiently
within Neovim.

Solution:
Add a DiffTool plugin and command that enables side-by-side diffing of
files and directories in Neovim. The plugin supports rename detection,
highlights changes in the quickfix list, and provides a user command for
easy invocation. This allows proper integration with git difftool -d for
directory comparison.

Example git config:

```ini
[diff]
    tool = nvim_difftool

[difftool "nvim_difftool"]
    cmd = nvim -c "packadd nvim.difftool" -c "DiffTool $LOCAL $REMOTE"
```

Example `git difftool -d master` result on this PR:

<img width="2544" height="1431" alt="image" src="https://github.com/user-attachments/assets/599fea1f-bf18-4010-a9ac-f02e38aa5728" />

The original implementation is mostly just grabbed from my code here: https://github.com/deathbeam/myplugins.nvim/blob/main/lua/myplugins/difftool.lua

Im not sure if command plugin is best route to go for this or if there is even interest for something like this in core (imo it makes a lot of sense in core because git difftool -d is very useful and this removes the need for most external diff plugins with proper diffmode config and usage, at least for me).

deathbeam committed:
feat(diff): add DiffTool command for file and directory comparison

Problem:
Neovim's built-in diff mode (nvim -d) does not support directory diffing
as required by git difftool -d. This makes it difficult to compare
entire directories, detect renames, and navigate changes efficiently
within Neovim.

Solution:
Add a DiffTool plugin and command that enables side-by-side diffing of
files and directories in Neovim. The plugin supports rename detection,
highlights changes in the quickfix list, and provides a user command for
easy invocation. This allows proper integration with git difftool -d for
directory comparison.

Example git config:

```ini
[diff]
    tool = nvim_difftool

[difftool "nvim_difftool"]
    cmd = nvim -c "packadd nvim.difftool" -c "DiffTool $LOCAL $REMOTE"
```

Co-authored-by: Phạm Bình An <111893501+brianhuster@users.noreply.github.com>
Co-authored-by: Justin M. Keyes <justinkz@gmail.com>
Signed-off-by: Tomas Slusny <slusnucky@gmail.com>

brianhuster commented on runtime/plugin/difftool.lua:
```diff
@@ -0,0 +1,12 @@
+  else
+    vim.notify('Usage: DiffTool <left> <right>', vim.log.levels.ERROR)
+  end
+end, { nargs = '*' })
```
```suggestion
end, { nargs = '*', complete = 'file' })
```

justinmk commented on runtime/plugin/difftool.lua:
```diff
@@ -0,0 +1,12 @@
+
+vim.api.nvim_create_user_command('DiffTool', function(opts)
+  if #opts.fargs == 2 then
+    require('difftool').diff(opts.fargs[1], opts.fargs[2])
```
Move `runtime/lua/difftool.lua` to `runtime/lua/nvim/difftool.lua` , that's where plugins live. This is not part of the stdlib.

```suggestion
    require('nvim.difftool').diff(opts.fargs[1], opts.fargs[2])
```

brianhuster commented:
I think these should be taken into accounts:
- Some kind of "ignore" option
- Speed. I just tried it on 2 directories, and it has made Nvim hanged since even before I sent the comment https://github.com/neovim/neovim/pull/35448#discussion_r2296692340. Meanwhile `diff -r` still gives me result instantly. Unix has `diff` CLI tool for comparing file and directories, and it is extremely fast, so some way to use it as the provider would be nice.

deathbeam commented:
> I think these should be taken into accounts:
> 
> * Some kind of "ignore" option
> * Speed. I just tried it on 2 directories, and it has made Nvim hang since even before I sent the comment [feat(diff): add DiffTool plugin for file and directory comparison #35448 (comment)](https://github.com/neovim/neovim/pull/35448#discussion_r2296692340). Meanwhile `diff -r` still gives me result instantly. Unix has `diff` CLI tool for comparing file and directories, and it is extremely fast, so some way to use it as the provider would be nice. Or I wonder if there is a existing library for comparing directories?

I bet this is caused by the file rename detection, which can definitely be optimized, my implementation is quite basic. Looking at some stuff, I can use `git diff --no-index --name-status` to do this in generic way using git, that is maybe way better.

Also for ignore, what do you mean by that one?



> Please update the description (and commit message) as a Problem/Solution statement.
> 
> I'm open to this (have been using the rather buggy https://github.com/will133/vim-dirdiff for a long time) but will wait for feedback from others.
> 
> Can you publish it as a plugin so that people can try it out for awhile?

I actually do have it published as a plugin here: https://github.com/deathbeam/myplugins.nvim?tab=readme-ov-file#difftool

And alright will update the description in abit

justinmk commented:
> Looking at some stuff, I can use `git diff --no-index --name-status` to do this in generic way using git

If this is going to use git-specific stuff then that starts a different conversation: either it needs to be shaped in a way that allows "other VCS" support in the future, or would need to decide to make it more clearly "git-only" by naming it `gitdifftool`.

I would wait until there is actually feedback indicating that we want this in core...

> I actually do have it published as a plugin here: https://github.com/deathbeam/myplugins.nvim?tab=readme-ov-file#difftool

I mean an isolated plugin :)

deathbeam commented:
> > Looking at some stuff, I can use `git diff --no-index --name-status` to do this in generic way using git
> 
> If this is going to use git-specific stuff then that starts a different conversation: either it needs to be shaped in a way that allows "other VCS" support in the future, or would need to decide to make it more clearly "git-only" by naming it `gitdifftool`.
> 
> I would wait until there is actually feedback indicating that we want this in core...
> 
> > I actually do have it published as a plugin here: https://github.com/deathbeam/myplugins.nvim?tab=readme-ov-file#difftool
> 
> I mean an isolated plugin :)

Its not really git specific, --no-index means it doesnt need git working tree at all, just git cli and can work on any directory. 

And sure I can extract it to isolated plugin repo

deathbeam commented:
Alright created separate repo here: https://github.com/deathbeam/difftool.nvim

Mostly with same implementation as here, and updated the commit message too

deathbeam commented:
> are there plans to allow this dir diff through the existing `nvim -d` interface? it would make sense to me to extend the existing diff feature instead of adding an interface through a command. it would also fix some bugs caused by the directories not being recognized as arguments to neovim.

I was thinking about this but I did not wanted to break existing functionality however unituitive it is for comparing directories and also biggest reason is that supporting more than 2 directories here is tricky (with `diff` provider also kinda impossible as it only supports 2) and nvim -d supports any amount of inputs afaik

justinmk commented:
Similar to https://github.com/neovim/neovim/pull/35627 , this should probably live in `runtime/pack/dist/opt/nvim.difftool` , at least initially.

deathbeam commented:
> > so it works out of the box for git
> 
> meaning what? how does git use it? and why can't it call `:packadd nvim.difftool`

means doing this:

cmd = nvim -c "DiffTool $LOCAL $REMOTE"

but i guess it can also just call packadd as well so its probably whatever

deathbeam commented:
For now moved the plugin to directory @justinmk suggested and updated the commit message. For replacing nvim -d, i guess the replacement is few lines of code yes but as I said earlier, nvim -d supports arbitrary number of arguments, adding support to arbitrary number of arguments here is not trivial (and with the `diff` backend not possible even). But maybe as optional feature it could be fine?

justinmk commented:
`nvim -d` is an interesting idea that we can explore in a later PR, after we think difftool is mature enough.


> does vim.fn.argv() somehow depend on buffer names

That gets the vim "argument list" which is different than the startup file args, it can be modified (renamed/added/removed) at any time by a plugin. `v:argv` has the raw args passed at startup (which will include non-file args), but that might be annoying to deal with since the last args are not always "file args" (e.g. `nvim -l ...`).

There is no reason for dir plugins like Oil to rename the buffers, that's pretty dumb.

justinmk reviewed (changes requested):
test

deathbeam commented:
Added tests and also added a bit more documentation + adjusted gen_vimdoc to generate docs for it

justinmk commented on runtime/doc/news.txt:
```diff
@@ -196,6 +196,7 @@ EDITOR
   "(v)iew" then run `:trust`.
 • |gx| in help buffers opens the online documentation for the tag under the
   cursor.
+• New |:DiffTool| command for comparing two files or directories side-by-side
```
```suggestion
• The |:DiffTool| command compares directories (and files).
```

justinmk commented on runtime/doc/plugins.txt:
```diff
@@ -157,4 +158,42 @@ tohtml({winid}, {opt})                                       *tohtml.tohtml()*
+
+
+:DiffTool {left} {right}                                           *:DiffTool*
+Compares two files or directories side-by-side in Neovim.
```
"in Neovim" is redundant. it's assumed.

justinmk commented on runtime/doc/plugins.txt:
```diff
@@ -157,4 +158,42 @@ tohtml({winid}, {opt})                                       *tohtml.tohtml()*
+
+
+close()                                                     *difftool.close()*
+    Close the diff layout and clean up
```
why is this needed? why can't it handle BufDelete and/or other events? this `close()` function is already implicitly assuming "the current buffer" so it's not even that useful anyway.

deathbeam replied:
hmm, i just had it there because i kinda liked just doing :DiffTool to also close but yea its not super useful and its better if it cleans up when one of the wins is closed

justinmk commented on src/gen/gen_vimdoc.lua:
```diff
@@ -415,10 +415,12 @@ local config = {
     section_order = {
       'editorconfig.lua',
       'tohtml.lua',
+      'difftool.lua',
```
alphanum order

justinmk reviewed (approved):
Once this matures maybe it should be builtin so that `:diffthis` works on directories :)

deathbeam commented:
> @deathbeam comparing 2 dirs that are identical seems to raise a "No errors" error (empty quickfix list I guess).
> 
> ```
> :DiffTool foo/ foo/
> ```
> 
> ```
> Lua :command callback: ...vim/runtime/pack/dist/opt/nvim.difftool/lua/difftool.lua:395: Vim:E42: No Errors
> stack traceback:
> 	[C]: in function 'cfirst'
> 	...vim/runtime/pack/dist/opt/nvim.difftool/lua/difftool.lua:395: in function 'diff_dirs'
> 	...vim/runtime/pack/dist/opt/nvim.difftool/lua/difftool.lua:504: in function 'open'
> 	.../runtime/pack/dist/opt/nvim.difftool/plugin/difftool.lua:8: in function <.../runtime/pack/dist/opt/nvim.difftool/plugin/difftool.lua:6>
> ```

Ah, good catch, opened PR with fix here (early return with vim.notify is best solution I think): https://github.com/neovim/neovim/pull/36671

justinmk commented:
re https://github.com/neovim/neovim/pull/35448#issuecomment-3315105002 @deathbeam , you can now use `v:argf` : https://github.com/neovim/neovim/issues/35847
